src/k8s_upgrade.py

```diff
@@ -0,0 +1,421 @@
+                message = "Highest number unit is unhealthy. Refresh will not resume."
+                logger.debug(f"Resume refresh event failed: {message}")
+                action_event.fail(message)
+                return
+            # If a unit was unhealthy and the upgrade was forced, only the next unit will
+            # upgrade. As long as 1 or more units are unhealthy, the upgrade will need to be
+            # forced for each unit.
+
+            # Include "Attempting to" because (on Kubernetes) we only control the partition,
+            # not which units upgrade. Kubernetes may not upgrade a unit even if the partition
+            # allows it (e.g. if the charm container of a higher unit is not ready). This is
+            # also applicable `if not force`, but is unlikely to happen since all units are
+            # healthy `if not force`.
+            message = f"Attempting to refresh unit {self._partition}."
+            action_event.set_results({"result": message})
+            logger.debug(f"Resume refresh succeeded: {message}")
+
+
+class _PostUpgradeCheckMongoDB(EventBase):
+    """Run post upgrade check on MongoDB to verify that the cluster is healhty."""
+
+    def __init__(self, handle):
+        super().__init__(handle)
+
+
+class MongosUpgrade(GenericMongosUpgrade):
+    """Handlers for upgrade events."""
+
+    post_app_upgrade_event = EventSource(_PostUpgradeCheckMongoDB)
+
+    def __init__(self, charm: "MongosCharm"):
+        self.charm = charm
+        super().__init__(charm, PEER_RELATION_ENDPOINT_NAME)
+
+    @override
+    def _observe_events(self, charm: "MongosCharm") -> None:
+        self.framework.observe(
+            charm.on[PEER_RELATION_ENDPOINT_NAME].relation_created,
+            self._on_upgrade_peer_relation_created,
+        )
+        self.framework.observe(
+            charm.on[PEER_RELATION_ENDPOINT_NAME].relation_changed,
+            self._reconcile_upgrade,
+        )
+        self.framework.observe(
+            charm.on[PRECHECK_ACTION_NAME].action, self._on_pre_upgrade_check_action
+        )
+        self.framework.observe(
+            self.post_app_upgrade_event, self.run_post_app_upgrade_task
+        )
+        self.framework.observe(
+            charm.on["force-refresh-start"].action, self._on_force_upgrade_action
+        )
+
+    def _on_force_upgrade_action(self, event: ActionEvent):
+        if not self.charm.unit.is_leader():
+            message = f"Must run action on leader unit. (e.g. `juju run {self.charm.app.name}/leader force-refresh-start`)"
+            logger.debug(f"Force refresh failed: {message}")
+            event.fail(message)
+            return
+        if not self._upgrade or not self._upgrade.in_progress:
+            message = "No upgrade in progress"
+            logger.debug(f"Force refresh failed: {message}")
+            event.fail(message)
+            return
+        self._upgrade.reconcile_partition(action_event=event)
```

Comment:
@Mehdi-Bendriss : This is mongos specific because we have a `force-refresh-start` event